src/data/polynomial/ring_division.lean

```diff
@@ -227,16 +230,62 @@ variables [comm_semiring R] [no_zero_divisors R] {p q : R[X]}
-        (or.inr ∘ is_unit_of_mul_eq_one g _) (or.inl ∘ is_unit_of_mul_eq_one f _)⟩⟩,
+  refine ⟨λ h f g hf hg hp, (h.2 f g hp.symm).imp hf.eq_one_of_is_unit hg.eq_one_of_is_unit,
+    λ h, ⟨hp1 ∘ hp.eq_one_of_is_unit, λ f g hfg, (h (g * C f.leading_coeff) (f * C g.leading_coeff)
+      _ _ _).symm.imp (is_unit_of_mul_eq_one f _) (is_unit_of_mul_eq_one g _)⟩⟩,
   { rwa [monic, leading_coeff_mul, leading_coeff_C, ←leading_coeff_mul, mul_comm, ←hfg, ←monic] },
   { rwa [monic, leading_coeff_mul, leading_coeff_C, ←leading_coeff_mul, ←hfg, ←monic] },
   { rw [mul_mul_mul_comm, ←C_mul, ←leading_coeff_mul, ←hfg, hp.leading_coeff, C_1, mul_one,
         mul_comm, ←hfg] },
 end
 
+lemma monic.irreducible_iff_nat_degree (hp : p.monic) : irreducible p ↔
+  p ≠ 1 ∧ ∀ f g : R[X], f.monic → g.monic → f * g = p → f.nat_degree = 0 ∨ g.nat_degree = 0 :=
+begin
+  by_cases hp1 : p = 1, { simp [hp1] },
+  rw [irreducible_of_monic hp hp1, and_iff_right hp1],
+  refine forall₄_congr (λ a b ha hb, _),
+  rw [ha.nat_degree_eq_zero_iff_eq_one, hb.nat_degree_eq_zero_iff_eq_one],
+end
+
+lemma monic.irreducible_iff_nat_degree' (hp : p.monic) : irreducible p ↔ p ≠ 1 ∧
+  ∀ f g : R[X], f.monic → g.monic → f * g = p → g.nat_degree ∉ finset.Ioc 0 (p.nat_degree / 2) :=
+begin
+  simp_rw [hp.irreducible_iff_nat_degree,
+    finset.mem_Ioc, nat.le_div_iff_mul_le zero_lt_two, mul_two],
+  apply and_congr iff.rfl,
+  split; intros h f g hf hg he; subst he,
+  { refine λ ha, (h g f hg hf $ mul_comm g f).elim ha.1.ne' _,
+    rw hf.nat_degree_mul hg at ha,
+    exact (ha.1.trans_le $ (add_le_add_iff_right _).1 ha.2).ne' },
+  { simp_rw hf.nat_degree_mul hg at h,
+    by_contra hdp, simp_rw [not_or_distrib, ← ne.def, ← pos_iff_ne_zero] at hdp,
+    obtain hl|hl := le_total f.nat_degree g.nat_degree,
+    { exact h g f hg hf (mul_comm g f) ⟨hdp.1, add_le_add_left hl _⟩ },
+    { exact h f g hf hg rfl ⟨hdp.2, add_le_add_right hl _⟩ } },
+end
+
+lemma monic.not_irreducible_iff_exists_add_mul_eq_coeff (hm : p.monic) (hnd : p.nat_degree = 2) :
+  ¬ irreducible p ↔ ∃ c₁ c₂, p.coeff 0 = c₁ * c₂ ∧ p.coeff 1 = c₁ + c₂ :=
+begin
+  casesI subsingleton_or_nontrivial R,
+  { simpa only [nat_degree_of_subsingleton] using hnd },
+  rw [hm.irreducible_iff_nat_degree', hnd],
+  push_neg,
+  rw imp_iff_right,
+  split,
+  { rintros ⟨a, b, ha, hb, rfl, hdb|⟨⟨⟩⟩⟩,
+    have hda := hnd, rw [ha.nat_degree_mul hb, hdb] at hda,
+    use [a.coeff 0, b.coeff 0, mul_coeff_zero a b],
+    simpa only [next_coeff, hnd, add_right_cancel hda, hdb] using ha.next_coeff_mul hb },
+  { rintros ⟨c₁, c₂, hmul, hadd⟩,
+    refine ⟨X + C c₁, X + C c₂, monic_X_add_C _, monic_X_add_C _, _, or.inl $ nat_degree_X_add_C _⟩,
+    rw [p.as_sum_range_C_mul_X_pow, hnd, finset.sum_range_succ, finset.sum_range_succ,
+      finset.sum_range_one, ← hnd, hm.coeff_nat_degree, hnd, hmul, hadd, C_mul, C_add, C_1],
+    ring },
+  { rintro rfl, simpa only [nat_degree_one] using hnd },
```

Comment:
This avoids the `hdb|⟨⟨⟩⟩` hack, which should allow you to switch to `set.Ico` (which I think is preferable).
```suggestion
  have hp1 : p ≠ 1 := λ hp1, by simpa only [hp1, nat_degree_one] using hnd,
  simp_rw [hm.irreducible_iff_nat_degree', not_and_distrib, not_forall, not_not, hp1, false_or,
    hnd, nat.div_self two_pos, nat.Ioc_succ_singleton, zero_add, mem_singleton],
  split,
  { rintros ⟨a, b, ha, hb, rfl, hdb⟩,
    have hda := hnd, rw [ha.nat_degree_mul hb, hdb] at hda,
    use [a.coeff 0, b.coeff 0, mul_coeff_zero a b],
    simpa only [next_coeff, hnd, add_right_cancel hda, hdb] using ha.next_coeff_mul hb },
  { rintros ⟨c₁, c₂, hmul, hadd⟩,
    refine ⟨X + C c₁, X + C c₂, monic_X_add_C _, monic_X_add_C _, _, nat_degree_X_add_C _⟩,
    rw [p.as_sum_range_C_mul_X_pow, hnd, finset.sum_range_succ, finset.sum_range_succ,
      finset.sum_range_one, ← hnd, hm.coeff_nat_degree, hnd, hmul, hadd, C_mul, C_add, C_1],
    ring },
```

Reply:
@tb65536 I'm not sure why you think `hdb|⟨⟨⟩⟩` is a hack to be avoided; isn't this how you're supposed to use `finset`, whose purpose is to include the data of an enumeration of its elements? Why should `finset.Ioc` exist as a (computable) def in mathlib if not for enumerating the elements? Why should we use all those lemmas to pin down the element when we have access to an enumeration? Even if the lemma is to be stated with `set.Ioc`, I think the proof should still use `hdb|⟨⟨⟩⟩` after converting to `finset.Ioc` using [finset.coe_Ioc](https://leanprover-community.github.io/mathlib_docs/order/locally_finite.html#finset.coe_Ioc).

I'm also not sure why you'd like to squeeze the `push_neg`; it seems to be pretty fast (and predictable) as it involves no branching.

Reply:
To me, I would rather use API. But maybe it's a matter of personal preference.

And I wasn't explicitly to squeeze the `push_neg`. I just thought that you might as well use `simp_rw` to simplify `g.nat_degree ∉ Ioc 0 (p.nat_degree / 2)` and `p ≠ 1` if you have the opportunity. And it seemed easier to manually push the negations than to try to break up the `simp_rw`-chain and insert a `push_neg`.